SHS-6316: Add DDEV support

nJim opened the pull request:
# Summary
This PR migrates the local development environment from Lando to DDEV, updates all documentation and scripts to use DDEV, and updates lando specific code to work concurrently for select users who chose not to migrate.

## Backstory
- Our agency is standardizing on DDEV for local development to improve compatibility, support, and onboarding. This work was tracked in SHS-6316.
- All documentation, theme scripts, and configuration files have been updated to reference DDEV instead of Lando.
- Lando-specific files and instructions have been preserved but are noted as a legacy tool.
- This ensures a consistent and modern local development experience for all contributors.
- Not included in this work: Setup of local SAML auth and setup of local CodeCeption. These tools were not correctly installing in Lando and have been documented in the DDEV readme as areas of future enhancement. 

## Need Review By (Date)
Not set.

## Urgency
Low.

## Steps to Test
- Prepare for testing: Checkout this branch and ensure you have DDEV installed and meet system requirements. Follow the updated .ddev/DDEV-README.md for local setup instructions. While not required, you may need to quit lando and possibly destroy those containers to feel confident in your testing.
```bash
# Reset localdev in case you have anything in the DDEV container already
ddev delete --omit-snapshot
# Start DDEV (optional step as the next command will do this too)
ddev start
# Verify containers are running (no surprises or errors)
ddev describe
# Create a local environment
ddev blt drupal:sync --site=hs_traditional
# Verify site initialization (no surprises or errors)
ddev drush @hs_traditional.local status
# Verify that the site exists and functional locally
open https://hs-traditional.ddev.site/
open http://hs-traditional.ddev.site/
# Create another one
ddev blt drupal:sync --site=art
# Verify that the site exists and functional locally
open https://art.ddev.site/
open http://art.ddev.site/
# Check that the first site is also still avilable
open https://hs-traditional.ddev.site/
open http://hs-traditional.ddev.site/
# Test a drush command to log in
ddev drush @hs_traditional.local uli
# Test a theme command
npm run theme-build
```

- Run all theme build, lint, and watch scripts (hooks/common/theme/) and confirm they function using DDEV.
- Validate the updated documentation. Verify that no Lando references remain in documentation, scripts, or config.
- Verify that the environment can still be run with lando.

# Review Tasks

## Backend / Functional Validation
### Code
- [ ] Are the naming conventions following our standards?
- [ ] Are PHP functions and variables in `snake_case` and not `camelCase`?
- [ ] Does Drupal code follow [Drupal Coding Standards](https://www.drupal.org/docs/develop/standards/php/php-coding-standards)?
- [ ] Does the code have sufficient inline comments?
- [ ] Is there anything in this code that would be hidden or hard to discover through the UI?
- [ ] Are there any [code smells](https://blog.codinghorror.com/code-smells/)?
- [ ] Are tests provided?

### Code security
- [ ] Is all [user input properly sanitized when rendered](https://www.drupal.org/docs/8/security/drupal-8-sanitizing-output)?
- [ ] Any obvious [security flaws or new areas for attack](https://www.drupal.org/docs/8/security)?

### General
- [ ] Is there anything included in this PR that is not related to the problem it is trying to solve?
- [ ] Is the approach to the problem appropriate?

nJim committed:
feat(SHS-6316): migration from lando to ddev for localdev

nJim committed:
feat(SHS-6316): prevent ddev updates to settings.php

nJim committed:
feat(SHS-6316): update ddev docs and scripts

nJim committed:
Merge branch '11.12.2-release' into SHS-6316-ddev

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1906-ib1licfdepkny1uq6f3argnarcze6rgk.tugboatqa.com
- https://hs-colorful-ib1licfdepkny1uq6f3argnarcze6rgk.tugboatqa.com
- https://hs-traditional-ib1licfdepkny1uq6f3argnarcze6rgk.tugboatqa.com
- https://english-ib1licfdepkny1uq6f3argnarcze6rgk.tugboatqa.com
- https://history-ib1licfdepkny1uq6f3argnarcze6rgk.tugboatqa.com
- https://west-ib1licfdepkny1uq6f3argnarcze6rgk.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/687e4cb1e0ac79bcee978bf6

Visual Diffs:
- https://dashboard.tugboatqa.com/687e4cb1e0ac79bcee978bf6/visualdiffs

nJim committed:
feat(SHS-6316): reinstate lando as legacy option

nJim commented:
@cienvaras Lando support added back to the project, although it is noted as a legacy tool in the main readme. Theme build scripts will check if the user is running ddev, then check if lando is running, or finally fallback for any other localdev. This is a continuation of the existing pattern.

cienvaras commented on lando/README.md:
````diff
@@ -128,7 +128,6 @@ If you want to use [Lando](https://lando.dev/) for local development, here are s
       127.0.0.1 womensleadership.suhumsci.loc
       127.0.0.1 womensleadershipcp.suhumsci.loc
     ```
-
````
Can you please restore the line breaks? It doesn't affect the markdown rendering, but makes the source file harder to read.

nJim replied:
This was accidentally removed when I was trying to restore this file. Fixed.

cienvaras reviewed (changes requested):
@nJim Looks really good! Definitely feels faster than lando, even using docker.

Just found an issue with the FE scripts. When running `ddev blt drupal:sync...`, the FE build script runs automatically at the end. It worked as expected with lando, but with ddev there's an error, I think it's related with the logic for finding the right npm command:

```
> source:build:frontend-reqs
Executing frontend-reqs target hook...
 [ExecStack] ./hooks/common/theme/theme-build.sh
 [ExecStack] Running ./hooks/common/theme/theme-build.sh in /var/www/html
Installing npm dependencies...
You executed 'ddev npm ci' inside the web container
but many DDEV commands are not available.

However, you may be able to use a command like:

npm ci
Building theme...
You executed 'ddev npm run build' inside the web container
but many DDEV commands are not available.

However, you may be able to use a command like:

npm run build
```

nJim committed:
feat(SHS-6316): run npm commands within docker container

nJim commented:
Thanks @cienvaras . I've updated the theme build scripts. This broke when reintroducing lando. Now I know that DDEV is a valid command in a DDEV managed container, but it only tells you not to use the command. So it was failing too gracefully. I've changed the logic to use just 'npm'. I've tested a DDEV and Lando build on my local.

cienvaras reviewed (changes requested):
@nJim Thanks for the fix, the FE scripts are working as expected now.

I'm having a problem though. I installed `hs_colorful`, tested it and everything worked as expected. Then, I installed `physics` and now https://hs-colorful.ddev.site/ (and all other domains) points to the physics site instead of the previously installed `hs_colorful`. In lando it was possible to have multiple sites installed at the same time, each one on its own domain. Is this a limitation of ddev or is something misconfigured?

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim commented:
@cienvaras This issue should be resolved. DDEV databases now have unique names so multiple can be run at the same time. This matches how things worked in lando.

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1906-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-colorful-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-traditional-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://english-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://history-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://west-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c

Visual Diffs:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c/visualdiffs

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
Merge branch '11.13.1-release' into SHS-6316-ddev

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1906-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-colorful-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-traditional-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://english-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://history-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://west-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c

Visual Diffs:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c/visualdiffs

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1906-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-colorful-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-traditional-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://english-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://history-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://west-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c

Visual Diffs:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c/visualdiffs

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1906-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-colorful-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://hs-traditional-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://english-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://history-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com
- https://west-cei8mk73sdxeao9w6fikereqtsg29egz.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c

Visual Diffs:
- https://dashboard.tugboatqa.com/6891f910735236489f16db2c/visualdiffs

nJim committed:
fix(SHS-6316): fix for unique database names in ddev

nJim committed:
Merge branch '11.14.1-release' into SHS-6316-ddev

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1906-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://hs-colorful-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://hs-traditional-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://english-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://history-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://west-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/68ab7bf954281768607a2add

Visual Diffs:
- https://dashboard.tugboatqa.com/68ab7bf954281768607a2add/visualdiffs

nJim committed:
fix(SHS-6316): allow internal path to ddev sites

ahughes3 commented:
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Links:
- https://pr1906-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://hs-colorful-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://hs-traditional-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://english-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://history-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com
- https://west-obobgh2gnckd2dznk9dksvmpu7nyrpv9.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/68ab7bf954281768607a2add

Visual Diffs:
- https://dashboard.tugboatqa.com/68ab7bf954281768607a2add/visualdiffs

nJim commented:
@cienvaras This is ready again for testing

cienvaras reviewed (approved):
@nJim Works like a charm, thanks for the hard work! :shipit:

cienvaras commented:
@Jnwibowo7321 @ahughes3 This is purely dev stuff, but still assigning to you in case you want to take a look. Feel free to pass it directly to SWS.

nJim committed:
Merge branch '11.15.1-release' into SHS-6316-ddev

Jnwibowo7321 reviewed (approved):
LGTM

joegl commented on .ddev/config.yaml:
```diff
@@ -0,0 +1,148 @@
+php_version: "8.3"
+webserver_type: nginx-fpm
+xdebug_enabled: false
+additional_hostnames:
```
Is there no way to dynamically pull these from the multsites array in `blt/blt.yml`? It's not a big deal but this list will have to be manually maintained as is. Perhaps there can be a script in DDEV that updates these on setup?

joegl commented on README.md:
```diff
@@ -11,7 +11,7 @@ Note the following properties of this project:
 * Four Kitchens secondary release branch naming convention: `fk-stnfd-sprint-[SPRINT_NUMBER]`
 * Development branching convention: branch off the current release branch
-* Local environment: Lando or DrupalVM
+* Local environment: DDEV, Lando or DrupalVM
```
Is there actually support for `DrupalVM` in this repo and/or are you using it? We could probably drop it if it's a legacy thing.

joegl commented on README.md:
```diff
@@ -21,10 +21,11 @@ Note the following properties of this project:
 
 ### Local setup and installation
-You can either run the site on Lando or bare metal.
+You can either run the site on DDEV, Lando or bare metal.
```
I think it's fair to say DDEV is more up to date and preferred at this point if you want to make a note of that.

joegl commented on hooks/common/theme/theme-build.sh:
```diff
@@ -1,8 +1,14 @@
-# Run scripts via lando if available
-if [ -e .lando.yml ] && command -v lando &> /dev/null; then
+if [ -n "$DDEV_PROJECT" ] || [ -n "$LANDO_PROJECT" ]; then
+  # If inside a container, use npm directly.
+  NPM_CMD="npm"
+elif [ -e .ddev/config.yaml ] && command -v ddev &> /dev/null; then
+  # If outside a DDEV container, use ddev npm.
+  NPM_CMD="ddev npm"
+elif [ -e .lando.yml ]  && command -v lando &> /dev/null; then
+  # If outside a Lando container, use lando npm.
```
Is the "inside/outside" terminology inversed in the comments here? Or am I reading it backwards.

Also, I wonder if we should create a single wrapper command we could run these with, and then wrapper would determine whether to run the command prepended with "ddev" or "lando".

joegl reviewed (approved):
Overall this LGTM and I am going to merge as is. I'm going to note the comments in this review for future enhancement.
